### ISSUE TITLE
point 'from' is not taken into account while creating route

### ISSUE THREAD

**Issue by icegood** [osmandapp/OsmAnd#17498]
### Description
point 'from' is not taken into account while creating a route
### Steps to reproduce
1. Create route, in 'from' field select any point via 'select on map'
2. Select any point into 'to' field
### Actual result
route 'from py position'-'to' is created if the 'my position' could be obtained via GPS, or hangs if not
### Expected result
route 'from'-'to' will be created
### Your Environment
OsmAnd Version:  4.4.7 on android
Android/iOS version: EMUI v 12.0.0
Device model: huawei p20 pro

**Comment by scaidermern** [comment]
Don't press start if you want to keep a self defined starting location. Only after pressing start, from is replaced with your current position (which makes sense).

Alternatively, use the "plan a route" tool.

**Comment by icegood** [comment]
Strange UI design. From UX point of view app shouldn't take into account current location unless it is explicitly pointed out.

**Comment by scaidermern** [comment]
Yes, no, maybe. I don't know. ~~Can you repeat the question?~~
How is active routing supposed to work with a location that isn't yours?

**Comment by zipav** [comment]
OsmAnd Nightly 4.5.0#296m, released: 2023-06-27

### Steps to reproduce

- Activate location access
- Navigation button
- From: Select on map
- Select the start point on map
- To: select on map
- Select the finish point on map
- Check the result
- Repeat with with location access disabled 

The route successfully changes the starting point to another one with both the location turned on and the location turned off

https://github.com/osmandapp/OsmAnd/assets/123936521/e0aa0f1f-6ad7-4646-92f0-b4222cd19408


https://github.com/osmandapp/OsmAnd/assets/123936521/5896f1b1-0348-411c-8536-edf034a92f4d

**Comment by sonora** [comment]
The UX/UI aspect is valid, but it needs to be worked out how it could be decisively improved. What in my experience is regularly not understood by users is the difference between showing a route vs. entering an actual turn-by-turn navigation, as follows:
- (1) The route is calculated and ultimately displayed after a destination has been specified (and **without (before)** the "Start" button is tapped!)
  - At this stage, any arbitrary "From" location is considered, as selected
- (2) Tapping the "Start" button is often misunderstood as starting the route calculation, but it actually starts the turn-by-turn navigation.
  - Changing into this turn-by-turn mode may happen before the route is calculated and displayed, somewhat obsuring this mode change.
   - The turn,by,turn mode of course always routes from your actual position, all else would not make sense.
   - While your position is still unknown (or if your GPS is off), the turn-by-turn navigation mode pauses until it becomes known and it can start

**Comment by mikehgentry** [comment]
If you select Navigation, and set a From (X) and To (Y) point, when you hit start you're navigated from your position to Y. The green 'Start' flag is still displayed at X, but ignored.

Perhaps this is the source of confusion - intuitively I would expect it to take me from my position to X then to Y (i.e. move the specified 'Start' position to 'First intermediate' in the list).

**Comment by mikehgentry** [comment]
Route from X is an explicitly chosen option, with route from my location as another available choice (and the default, if nothing is chosen). There are situations where you genuinely want it to apply (I want to follow this nice route down this river, start at this end and finish at the other), and I can't see an obvious way someone would choose it accidentally (what would they be expecting to happen instead?).

**Comment by scaidermern** [comment]
OsmAnd could ask what to do when pressing start and a custom "from" has been set:
 * Replace from with the current position
 * Use from as the first via point

However I'm not sure if this is really a good idea. What do you think?

**Comment by mikehgentry** [comment]
I suppose there are more sensible options that could be added - 'route to the start', 'route to the nearest point', and 'route to the end' all seem fairly obvious, and I'm sure there are others. 

So personally I don't think having a dialogue box does much harm, even if it is arguably redundant.

**Comment by pordini** [comment]
Wouldn't routing from "my position" to "X" just be the same as setting a "first intermediate destination"?

Today I had a route suggesting an inaccessible forest path for kilometres, while the main road was 10 metres away. In this situation, while in turn by turn, I would like to select a point near me on the main road, choose "Navigate", choose "Make this the point of departure" and expect osmand to route from that point, not from my current position.
Current behaviour was that this point of departure is completely ignored, which does not make a lot of sense.

That said, if Osmand would just route from my custom point of departure and ignore my position, as soon as I start moving it would trigger a route recalculation anyway, right? 
Or, if you suppress that, there would just be a static route and no instructions which is odd as well.
Or Osmand could start giving turn by turn instructions once I have found my way to the calculated route, but that may be confusing and not expected by most users. Plus that isn't useful at all if I get lost on my way towards the route.

**Comment by sonora** [comment]
In my mind, showing and announcing upcoming turns is always relative to **where you are**. I cannot think of any other justified use case(?).

Specifying a "From:" location, correspondingly, is always a hypothetical/informational action to inspect how to get from arbitrary A to B, and/or determine travel time needed for that.

Mixing the two modes, like specifying a "From:" location but then expecting OsmAnd to use that as an initial "To:" location, seems debatable. For that use case there is the explicit "Via:" (intermediate destination") mechanism.

In summary, I feel it would be best to simply make it clearer when you are in the mere "Show a route from A to B" case vs. when you have switched over to the "Actively guide me to a destination" mode?

**Comment by pordini** [comment]
I think I agree. 
And my use case was (at least for me) so obscure I could entirely live without it.

But the option to set a new departure point exists while in turn by turn and it does nothing besides putting a flag on the map. I think that's just a map marker, not an option that should be under Navigation. 
In my view, any option should not add confusion, especially when easily accessed while driving.
I think setting a departure point could just be hidden when in "Actively guide me to a destination" mode. Especially if, like now, it does not alter the route anyway.

**Comment by mikehgentry** [comment]
Right, but you've still got the facility to find a route between point X and Y for informational purposes, just by not hitting Start. Once you've hit Start you've entered turn by turn mode, so the question is whether it makes more sense to drop the previous start point (which seems pretty arbitrary when it's been explicitly entered as part of the route) or incorporate it. I can see users saying "I want to follow the route from X to Y", hitting Start, and being confused about why X doesn't feature. On the other hand I can't see any reason you would input an X and want something else to happen once you hit Start.

**Comment by DmitryAlexei** [comment]
- You can also save your route as a GPX track using *Details* section in navigation mode. 
https://osmand.net/docs/user/navigation/setup/route-details#share--export-actions

- After that, you can use the created GPX track for navigation and choose whether to start from the beginning of the track or its closest point.
https://osmand.net/docs/user/navigation/setup/gpx-navigation#user-start--finish--points

**Comment by sonora** [comment]
In summary, I can think of 3 potential actions we take:
- [ ] When in navigation mode, make sure that everywhere the "From:" location displays as "My position", and is not tappable/changeable.
- [ ] When on the route config screen and someone has selected a "From:" location, and then taps "Start", we could inform in a short toast message "Navigation guidance has been started, the "From:" location is being replaced by the current position".
- [ ] We could potentially enhance the caption of the "Start" button to say something like "Start navigating"?

**Comment by pordini** [comment]
The first seems like a good solution. 

Second point: a toast might easily be missed, I think. Maybe a screen bottom dialog would be clearer, like the confirmation dialog when ending navigation. 
Possibly with a timer, defaulting to current position to X, and an option to forget X and just start towards the destination? But that would introduce a new mechanic (dialog with countdown), idk if that is too confusing. 

Regarding the last point: that may be a good idea anyway, to really distinguish between "Plan route" and "Start navigating".
Maybe the caption could change, something like "Calculating" while the button is serving as the progress bar, then "Start navigating" once the route has been completed. 

I would also say that it may be slightly incongruous that the voice announces the trip stats after pressing start. To me, that information is part of the planning stage, so it adds to the diffusion of those two stages if it's only said in the navigating stage. Not a big deal though.

**Comment by Yogurt4** [comment]
It's not only an UX design flaw, I think.
Just think of a situation when there's no route between your current position and the desired start. (Like being on an island.) There will be no Start then.

The use case is simply to prepare a routing and start it later.

If one chooses a starting point, the routing must start from that point and not an (implicit/hidden) current position. Changing the UX label is only a sham.

**Comment by sonora** [comment]
>The use case is simply to prepare a routing and start it later.
- You can prepare the "From - To" route to view it, and leave it displayed on screen while still on the island, if you want.
- You can also use the "Stop" button to interim hide it again. And later call it up again under "Previous Route", then start it.
- But when you tap the "Start" button, that starts the turn-by-turn guidance, necessarily **from your position**.

Not sure what you expect changing about this, unless the UI.

**Comment by hman2** [comment]
"(which makes sense)" NO. It does not make any sense at all. A start point was given, then this (and nothing else) is the start point.

If you always calculate from the current position, then why do you ask the user for a start point at all? Completely superfluous IMHO.

**Comment by hman2** [comment]
To give you an example why a user would want to start from a favorite point instead of the current position: In Germany handling your phone while driving is illegal, because it is dangerous. If you plan to drive to a destination you know, but from there, you do not know he route by heart, you will calculate the route from a favorite, so everything is set up in advance and you can in all calmliness study the calculated route for known trouble spots or imperfections, which you couldn't do while driving.

**Comment by hman2** [comment]
Another reason to calculate between favorites: Sometimes the current position is not yet known, giving your surroundings (inside a building for instance) it might take a couple of minutes. Forcing users to start calculation from current instead of the favorite that a user has stated he or she **wants** to be the start point delays the whole usage of Osmand by these minutes! Frustrating...

**Comment by scaidermern** [comment]
All of this is already possible by setting a custom start point and _not pressing start_. Probably not intuitive enough, I agree.

**Comment by sonora** [comment]
- You can **calculate** a route from anywhere to anywhere, to view it, explore it, or save it for later use, e.g. in your OsmAnd history or as a file.
- You can **simulate** a navigation from anywhere to anywhere, i.e. ask the system to pretend mock locations like following such a route, to check out the turns and turn instructions.
- But you cannot **navigate** from an arbitrary location, navigation here means the process of guiding you from your actual location.

**Comment by hman2** [comment]
a) Starting from a favorite is necessary even if that in fact IS the current position. Say, if you are in a parking garage you won't have GPS, so it is essential to select a favorite so navigation can start immediately and does not have to wait for GPS coords of the current position.
b) If a user chooses a favorite that differs (more than a couple of meters) from current position, IMHO that is the (!) opportunity for an error message. Or rather, a dialog that asks the user "given start point differs from current position. Change start point into first waypoint Y/N?".

**Comment by scaidermern** [comment]
> Or rather, a dialog that asks the user "given start point differs from current position. Change start point into first waypoint Y/N?".
> And for that, OSMand should not wait for current position. As I wrote, this can take long, seemingly neverending minutes depending on GPS receiption and antenna gain etc.

A dialog with a question whether to replace the chosen start with the current position is probably a good idea. It avoids users repeatedly stumbling into this situation. Even if an active navigation with a 'wrong' starting point doesn't make much sense.

**Comment by sonora** [comment]
So we have 2 cases:

a) The parking garage case, where initially you have no GPS. Unless we broke something, this should work. OsmAnd produces the route and also produces the first direction prompt from your manually selected start point. (Of course there can be no update until there is a position lock, how could there be. But as soon as there is, it should work? Please try if you have a real life use case, just to check if this perhaps became broken.)

b) That case I don't understand. OsmAnd creates the route correctly, also from your arbitrarily selected start point (including the first direction), and switches to the live position as soon as it becomes known. What more do you expect, and what practical effect or use should error messages or dialogs have at this point?

It seems to me @hman2 is under the impression there can be ongoing usable navigation prompts for a an arbitrary place where you are not at? But how can there be, other than in a simulation case?

**Comment by hman2** [comment]
"a) The parking garage case, where initially you have no GPS. Unless we broke something, this should work. OsmAnd produces the route". Wrong. Osmand does nothing but producing the error message "current position unknown". No route is produced.
"b) That case I don't understand. OsmAnd creates the route correctly, also from your arbitrarily selected start point". Wrong. Osmand produces a route from the current position completely ignoring the entered start point.

**Comment by scaidermern** [comment]
@hman2 The assumption for both cases was to set a custom starting point.

**Comment by hman2** [comment]
Yes. But OSMand ignores that and (tries to) use the current position, If that is unavailable,  OSMand will wait for that a looooong time.

**Comment by scaidermern** [comment]
No, that's not correct.

In case a) OsmAnd will calculate a route from the custom starting point and keep it until a position is known. Then it will replace the custom position with your actual position and, if necessary, will recalculate the route. Perfect for a parking garage.

It is still unclear how b) is supposed to work according to you.

**Comment by sonora** [comment]
Related: #22595.
